ntoskrnl/io/iomgr/deviface.c

```diff
@@ -26,192 +27,542 @@ C_ASSERT(sizeof(L"{01234567-89ab-cdef-0123-456789abcdef}") == GUID_STRING_BYTES
+        PrefixString->MaximumLength = PrefixString->Length;
+
+        DPRINT("Prefix string is '%wZ'\n", PrefixString);
     }
 
-    RtlAppendUnicodeStringToString(&SubKeyName,
-                                   SymbolicLinkName);
+    if (MungedString)
+        *MungedString = MungedStringReal;
 
-    SubKeyName.Buffer[SubKeyName.Length / sizeof(WCHAR)] = UNICODE_NULL;
+    if (GuidString)
+        *GuidString = GuidStringReal;
 
-    SubKeyName.Buffer[0] = L'#';
-    SubKeyName.Buffer[1] = L'#';
-    SubKeyName.Buffer[2] = L'?';
-    SubKeyName.Buffer[3] = L'#';
+    if (ReferenceString)
+        *ReferenceString = ReferenceStringReal;
 
-    ReferenceString.Buffer = wcsrchr(SubKeyName.Buffer, '\\');
-    if (ReferenceString.Buffer != NULL)
+    if (ReferenceStringPresent)
+        *ReferenceStringPresent = ReferenceStringReal.Length > sizeof(WCHAR);
+
+    if (InterfaceClassGuid)
     {
-        ReferenceString.Buffer[0] = L'#';
+        /* Convert GUID string into a GUID and store it also */
+        Status = RtlGUIDFromString(&GuidStringReal, InterfaceClassGuid);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlGUIDFromString() failed, Status 0x%08lx\n", Status);
+            RtlFreeUnicodeString(&MungedStringReal);
+            RtlInitEmptyUnicodeString(MungedString, NULL, 0);
+        }
+    }
 
-        SubKeyName.Length = (USHORT)((ULONG_PTR)(ReferenceString.Buffer) - (ULONG_PTR)SubKeyName.Buffer);
-        ReferenceString.Length = SymbolicLinkName->Length - SubKeyName.Length;
+    /* We're done */
+    return Status;
+}
+
+/**
+ * @brief
+ * Retrieves a handles to the device and instance registry keys
+ * for the previously opened registry key handle of the specified symbolic link.
+ **/
+static
+NTSTATUS
+IopOpenOrCreateSymbolicLinkSubKeys(
+    _Out_opt_ PHANDLE DeviceHandle,
+    _Out_opt_ PULONG DeviceDisposition,
+    _Out_opt_ PHANDLE InstanceHandle,
+    _Out_opt_ PULONG InstanceDisposition,
+    _In_ HANDLE ClassHandle,
+    _In_ PCUNICODE_STRING SymbolicLinkName,
+    _In_ ACCESS_MASK DesiredAccess,
+    _In_ BOOLEAN Create)
+{
+    UNICODE_STRING ReferenceString = {0};
+    UNICODE_STRING SymbolicLink = {0};
+    HANDLE DeviceKeyHandle, InstanceKeyHandle;
+    ULONG DeviceKeyDisposition, InstanceKeyDisposition;
+    BOOLEAN ReferenceStringPresent = FALSE; /* Assuming no ref string by default */
+    NTSTATUS Status;
+    WCHAR Buffer;
+
+    DeviceKeyHandle = InstanceKeyHandle = NULL;
+
+    /* Build up new symbolic link */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength = SymbolicLinkName->Length;
+    SymbolicLink.Buffer = ExAllocatePoolWithTag(PagedPool, SymbolicLink.MaximumLength, TAG_IO);
+    if (!SymbolicLink.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Quit;
+    }
+
+    /* Copy the contents of an existing symbolic link */
+    RtlCopyUnicodeString(&SymbolicLink, SymbolicLinkName);
```

Comment:
```suggestion
    /* Duplicate the symbolic link (we'll modify it later) */
    Status = RtlDuplicateUnicodeString(0, SymbolicLinkName, &SymbolicLink);
    if (!NT_SUCCESS(Status))
    {
        DPRINT1("RtlDuplicateUnicodeString() failed\n");
        goto Quit;
    }
```

Reply:
Committed, thanks. Also added the status returned from `RtlDuplicateUnicodeString` into `DPRINT1`. :slightly_smiling_face: